File: server/src/database/database_manager.py
import os
import json
from typing import List
from fastapi import HTTPException
from sqlmodel import SQLModel, create_engine
from sqlalchemy.engine.base import Engine
from sqlalchemy.orm import Session

from utils.assets_utils import get_assets_path
from aif_types.common import RequestFileInfo
from aif_types.embeddings import EmbeddingMetadata
from aif_types.agents import AgentMetadata, UpdateAgentRequest, CreateOrUpdateAgentResponse
from aif_types.functions import FunctionMetadata, UpdateFunctionRequest, CreateOrUpdateFunctionResponse, DeleteFunctionResponse
from aif_types.chat import ChatHistory, ChatHistoryMessage, ChatRole


class DatabaseManager:
    _engine: Engine | None = None

    def __init__(self) -> None:
        file_name = os.environ.get("SQLITE_FILE_NAME")
        assets_path = get_assets_path()
        database_uri = f"sqlite:///{assets_path}/{file_name}"
        self._engine = create_engine(database_uri)
        SQLModel.metadata.create_all(self._engine)

    def save_db_model(self, db_model: SQLModel):
        with Session(self._engine) as session:
            session.add(db_model)
            session.commit()

    def update_embeddings_name(self, asset_id: str, name: str):
        with Session(self._engine) as session:
            embedding = session.get(EmbeddingMetadata, asset_id)
            if embedding is None:
                raise Exception(f"Embedding with id {id} not found")
            embedding.name = name
            session.commit()

    def load_embeddings_metadata(self, asset_id: str) -> EmbeddingMetadata | None:
        with Session(self._engine) as session:
            return session.get(EmbeddingMetadata, asset_id)

    def list_embeddings_metadata(self) -> List[EmbeddingMetadata]:
        with Session(self._engine) as session:
            return session.query(EmbeddingMetadata).all()

    def delete_embeddings_metadata(self, asset_id: str):
        with Session(self._engine) as session:
            embedding_metadata = session.get(EmbeddingMetadata, asset_id)
            session.delete(embedding_metadata)
            session.commit()

    def list_agents(self) -> List[AgentMetadata]:
        with Session(self._engine) as session:
            return session.query(AgentMetadata).all()

    # def load_model(self, id: str) -> AgentMetadata | None:
    #     with Session(self._engine) as session:
    #         return session.get(AgentMetadata, id)
        
    def update_agent(self, id: str, request: UpdateAgentRequest) -> CreateOrUpdateAgentResponse:
        with Session(self._engine) as session:
            agent = session.get(AgentMetadata, id)
            if agent is None:
                raise Exception(f"Agent with id {id} not found")

            agent.base_model_uri = request.base_model_uri if request.base_model_uri else agent.base_model_uri
            agent.name = request.name if request.name else agent.name
            # System prompt can be None
            agent.system_prompt = request.system_prompt if request.system_prompt is not None else agent.system_prompt
            agent.rag_asset_ids = request.rag_asset_ids if request.rag_asset_ids else agent.rag_asset_ids
            agent.function_asset_ids = request.function_asset_ids if request.function_asset_ids else agent.function_asset_ids

            session.commit()
            return CreateOrUpdateAgentResponse(agent_uri=agent.agent_uri)

    def delete_agent(self, id: str):
        with Session(self._engine) as session:
            agent_metadata = session.get(AgentMetadata, id)
            if agent_metadata is None:
                raise Exception(f"Agent with id {id} not found")
            session.delete(agent_metadata)
            session.commit()


    def add_chat_message(self, aif_agent_uri: str, id: str, role: ChatRole, content: str, files: List[RequestFileInfo] = []):
        with Session(self._engine) as session:
            chat_history = session.get(ChatHistory, id)
            if chat_history is None:
                chat_history = ChatHistory(id=id, aif_agent_uri=aif_agent_uri, messages="[]")
                session.add(chat_history)

            chatHistoryMessage = ChatHistoryMessage(role=role.name, content=content, files=files)
            messages_json = json.loads(chat_history.messages)
            messages_json.append(chatHistoryMessage.model_dump_json())
            chat_history.messages = json.dumps(messages_json)
            session.commit()


    def get_chat_history(self, id: str) -> List[ChatHistoryMessage] | None:
        with Session(self._engine) as session:
            return session.get(ChatHistory, id)


    def get_chat_history_messages(self, id: str) -> List[ChatHistoryMessage] | None:
        with Session(self._engine) as session:
            chat_history = session.get(ChatHistory, id)
            if chat_history is None:
                return None

            messages_json = json.loads(chat_history.messages) if chat_history.messages else []
            messages = []
            for message in messages_json:
                chatHistoryMessageJson = json.loads(message)
                files = [RequestFileInfo(**file) for file in chatHistoryMessageJson["files"]]
                chatHistoryMessage = ChatHistoryMessage(role=chatHistoryMessageJson["role"], content=chatHistoryMessageJson["content"], files=files)
                messages.append(chatHistoryMessage)
            return messages


    def delete_chat_history(self, id: str):
        with Session(self._engine) as session:
            chat_history = session.get(ChatHistory, id)
            if chat_history is not None:
                session.delete(chat_history)
                session.commit()


    def list_functions(self) -> List[FunctionMetadata]:
        with Session(self._engine) as session:
            return session.query(FunctionMetadata).all()


    def get_function(self, id: str) -> FunctionMetadata | None:
        with Session(self._engine) as session:
            return session.get(FunctionMetadata, id)


    def update_function(self, request: UpdateFunctionRequest) -> CreateOrUpdateFunctionResponse:
        with Session(self._engine) as session:
            function = session.get(FunctionMetadata, request.id)
            if function is None:
                raise HTTPException(status_code=404, detail=f"Function with id {request.id} not found")

            function.name = request.name if request.name else function.name
            session.commit()
            return CreateOrUpdateFunctionResponse(
                id=function.id,
                uri=function.uri,
                name=function.name,
                functions_path=function.functions_path,
                functions_name=function.functions_name,
            )


    def delete_function(self, id: str):
        with Session(self._engine) as session:
            function = session.get(FunctionMetadata, id)
            if function is None:
                raise HTTPException(status_code=404, detail=f"Function with id {id} not found")

            session.delete(function)
            session.commit()
            return DeleteFunctionResponse(id=id)

File: packages/server-js/src/database/DatabaseManager.ts
import * as fs from 'fs';
import * as path from 'path';
import { types } from "aifoundry-vscode-shared";
import AssetUtils from '../utils/assetUtils';
import { HttpException } from '../exceptions';


class DatabaseManager {
    private _databaseFolderPath: string;

    constructor() {
    }

    public setup(databaseName: string) {
        try {
            this._setupDataSource(databaseName);
        } catch (error) {
            throw error;
        }
    }

    // Shared ------------------------------------------------------------------

    public saveDbEntity(dbEntity: types.database.BaseEntity): types.database.BaseEntity {
        const fileName = this._getDatabaseEntityFilePath(dbEntity.entityName);
        let content = {};
        if (!fs.existsSync(fileName)) {
            fs.writeFileSync(fileName, JSON.stringify(content));
        } else {
            content = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        }

        content[dbEntity.id] = dbEntity;
        fs.writeFileSync(fileName, JSON.stringify(content));
        return dbEntity;
    }

    public getDbEntity(dbName: string, id: string): types.database.BaseEntity | null {
        const fileName = this._getDatabaseEntityFilePath(dbName);
        let content = {};
        if (!fs.existsSync(fileName)) {
            return null;
        } else {
            content = JSON.parse(fs.readFileSync(fileName, 'utf8'));
            return content[id] ?? null;
        }
    }

    public deleteDbEntity(dbName: string, id: string): boolean {
        const fileName = this._getDatabaseEntityFilePath(dbName);
        let content = {};
        if (!fs.existsSync(fileName)) {
            return false;
        } else {
            content = JSON.parse(fs.readFileSync(fileName, 'utf8'));
            if (content[id]) {
                delete content[id];
                fs.writeFileSync(fileName, JSON.stringify(content));
            } else {
                return false;
            }
            return true;
        }
    }

    public listDbEntities(dbName: string): types.database.BaseEntity[] {
        const fileName = this._getDatabaseEntityFilePath(dbName);
        let content = {};
        if (!fs.existsSync(fileName)) {
            fs.writeFileSync(fileName, JSON.stringify(content));
        } else {
            content = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        }
        return Object.values(content);
    }

    // LmProvider --------------------------------------------------------------

    public listLmProviderInfo(): types.database.LmProviderInfo[] {
        return this.listDbEntities(types.database.LmProviderInfo.ENTITY_NAME) as types.database.LmProviderInfo[];
    }

    public getLmProviderInfo(providerId: string): types.database.LmProviderInfo | null {
        return this.getDbEntity(types.database.LmProviderInfo.ENTITY_NAME, providerId) as types.database.LmProviderInfo;
    }

    // Embeddings --------------------------------------------------------------

    public saveEmbeddingsMetadata(embeddingMetadata: types.database.EmbeddingMetadata) {
        this.saveDbEntity(embeddingMetadata);
    }

    public listEmbeddingsMetadata() {
        return this.listDbEntities(types.database.EmbeddingMetadata.name) as types.database.EmbeddingMetadata[];
    }

    public getEmbeddingsMetadata(assetId: string): types.database.EmbeddingMetadata | null {
        return this.getDbEntity(types.database.EmbeddingMetadata.name, assetId) as types.database.EmbeddingMetadata;
    }

    public deleteEmbeddingsMetadata(assetId: string) {
        return this.deleteDbEntity(types.database.EmbeddingMetadata.name, assetId);
    }

    // Agents ------------------------------------------------------------------

    public listAgents() {
        return this.listDbEntities(types.database.AgentMetadata.name) as types.database.AgentMetadata[];
    }

    public getAgent(agentId: string): types.database.AgentMetadata | null {
        return this.getDbEntity(types.database.AgentMetadata.name, agentId) as types.database.AgentMetadata;
    }

    public updateAgent(agentId: string, request: types.api.UpdateAgentRequest): types.database.AgentMetadata {
        const agent = this.getAgent(agentId);
        if (!agent) {
            throw new HttpException(404, `Agent not found`);
        }

        agent.basemodelUri = request.basemodelUri || agent.basemodelUri;
        agent.name = request.name || agent.name;
        agent.systemPrompt = request.systemPrompt || agent.systemPrompt;
        agent.ragAssetIds = request.ragAssetIds || agent.ragAssetIds;
        agent.functionAssetIds = request.functionAssetIds || agent.functionAssetIds;
        return this.saveDbEntity(agent) as types.database.AgentMetadata;
    }

    public deleteAgent(id: string) {
        return this.deleteDbEntity(types.database.AgentMetadata.name, id);
    }

    // Chat --------------------------------------------------------------------

    public getChatHistory(sessionId: string): types.database.ChatHistory | null {
        return this.getDbEntity(types.database.ChatHistory.name, sessionId) as types.database.ChatHistory;
    }

    public addChatMessage(
        sessionId: string,
        aifAgentUri: string,
        role: types.api.ChatRole,
        content: types.database.ChatHistoryMessageContent,
        contentTextFormat: types.api.TextFormat,
    ) {
        let chatHistory = this.getChatHistory(sessionId);
        if (!chatHistory) {
            chatHistory = new types.database.ChatHistory(sessionId, aifAgentUri, []);
        }

        const chatHistoryMessage: types.database.ChatHistoryMessage = { role, content, contentTextFormat };
        (chatHistory.messages as types.database.ChatHistoryMessage[]).push(chatHistoryMessage);
        this.saveDbEntity(chatHistory);
    }


    // Private -----------------------------------------------------------------

    private _setupDataSource(databaseName: string) {
        const assetsPath = AssetUtils.getAssetsPath();
        this._databaseFolderPath = path.join(assetsPath, databaseName);

        // In solution for JSON based database, databaseName is the folder for all the json files
        if (!fs.existsSync(this._databaseFolderPath)) {
            fs.mkdirSync(this._databaseFolderPath, { recursive: true });
        }
    }

    private _getDatabaseEntityFilePath(dbName: string) {
        return path.join(this._databaseFolderPath, dbName + '.json');
    }
}

export default DatabaseManager;
